### DISCUSSION TITLE
Hosting service / multitenant capability?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#498] [Q&A]
@ganigeorgiev This thread is to capture discussion about how a hosting service might work. I've been experimenting with Dockerizing a build of PocketBase behind nginx for hosting on fly.io.

My idea is to create a service where anyone can spin up a PocketBase instance and get a URL like `<subdomain>.pockethost.io` and the connect to it from the JS/Dart client.

In thinking about this, I am curious to know how feasible you think it would be for PocketBase to be modified for multitenancy based on subdomain. Off the top of my head, I can imagine a few approaches:

1. Run a separate pocketbase instance for each active subdomain - I have not tested this, but I imagine dozens or hundreds of `pocketbase` processes would not play well together
2. Modify PocketBase for multitenancy, either by switching between SQLite instances on a per-request basis, or perhaps by adding a subdomain prefix to table names.

Anyway, just opening this for thoughts and feedback from the community. I'm very stoked about contributing.

**Comment by ganigeorgiev** [comment]
Personally I would go with 1 - have one "supervisor" app that takes care for spinning individual vms for each subdomain.

The second approach requires too much changes to the internals in order to be implemented properly.

The problem with this type of hosting is that your users will not be able to execute custom code (unless you allow them to upload custom executable) because PocketBase doesn't support cloud functions and custom code is expected to be writted by embedding PocketBase in the user's Go application(s).

To be honest, I don't fully understand why so many users seems to avoid self-hosting, because PocketBase was created exactly to make it easy to self host (see [Presentator #183](https://github.com/presentator/presentator/issues/183)). You basically need only 2 things:

1. Deploy the binary on a brand new VM

2. Create a systemd service to keep the process up:

```nginx
# vi /lib/systemd/system/pocketbase.service

[Unit]
Description = pocketbase

[Service]
Type             = simple
User             = root
Group            = root
LimitNOFILE      = 4096
Restart          = always
RestartSec       = 5s
StandardOutput   = append:/root/pb/errors.log
StandardError    = append:/root/pb/errors.log
ExecStart        = /root/pb/pocketbase serve --http="yourdomain.com:80" --https="yourdomain.com:443"

[Install]
WantedBy = multi-user.target
```

And then just enable the service, aka. `systemctl enable pocketbase.service --now`.

**Reply by denisgurev** [reply]
I've done exactly this for my Hetzner vps with their dns editor which took 5 mins probably (I'm using the cheapest plan). The only other thing I had to create was an account in sendinblue for sending emails but they have free options so for now I'm good. I'm in love with PocketBase ❤️

**Reply by osseonews** [reply]
"To be honest, I don't fully understand why so many users seems to avoid self-hosting..."
B/c for most people everything you wrote is simply unintelligible. I can program, but I can't for the life of me figure out how to set up a server to run this. Just don't understand any of the terminology whatsoever. "Deploy the binary on a brand new VM" - what is the binary, and how do you deploy it, e.g. to a Digital Ocean droplet? I've been able to use docker for some apps, so if this is dockerized that would help somewhat.  But, still hosting these things are beyond the capabilities of most people. Most people just want to be up and running and don't want to worry about setting anything up - and they will pay for it. Charge $25 a month for this and you will get tons of users.

**Reply by benallfree** [reply]
Maybe the hosted service should be called PocketBaaS 🗡️ 😆 

I wonder how many users actually need to customize PocketBase, or if the use cases are common enough to be abstracted to a BaaS dashboard. Do you have any insight into how PocketBase is being customized by the community?

> The second approach requires too much changes to the internals in order to be implemented properly.

I'm looking at that now. In the mean time, I'll see how many `pocketbase` processes a VM can handle :)

**Reply by benallfree** [reply]
@osseonews Please get in touch with me if you'd like to be a beta tester, I'm putting something together. Gani is working really hard on PocketBase and it can be frustrating to come in not knowing any of those devops terms. Even though I understand all the terms, I still don't *want* to set it up. I guess Supabase and Firebase spoiled me.

**Reply by ganigeorgiev** [reply]
@osseonews

> "Deploy the binary on a brand new VM" - what is the binary, and how do you deploy it, e.g. to a Digital Ocean droplet?

Yes, the VM (VPS) equivalent in DigitalOcean terms is "Droplet". You don't need to install anything in addition. Create for example a Ubuntu (or any other linux) droplet. [Download and extract the zip from the releases page for linux amd64](https://github.com/pocketbase/pocketbase/releases) and there you'll find a `pocketbase` binary. It is enough to upload this binary on the DigitalOcean droplet and start it as in the example with systemd service. And that's it.

If you need further help, feel free to start a new discussion. The documentation will be refactored in the near future and I'll have it in mind to simplify the "Going to production" guide.

**Reply by ganigeorgiev** [reply]
@benallfree 

> I wonder how many users actually need to customize PocketBase, or if the use cases are common enough to be abstracted to a BaaS dashboard. Do you have any insight into how PocketBase is being customized by the community?

I don't really know, but at some point most likely you'll need to execute custom code. And to handle this usually you'll create a middleware service or use cloud functions (in the case of PocketBase - embed in a Go application). PocketBase was created to assist me developing the next version of my other open source project, Presentator (you could read more why in [Presentator #183](https://github.com/presentator/presentator/issues/183)).

**Reply by tonyhart7** [reply]
> "To be honest, I don't fully understand why so many users seems to avoid self-hosting..." B/c for most people everything you wrote is simply unintelligible. I can program, but I can't for the life of me figure out how to set up a server to run this. Just don't understand any of the terminology whatsoever. "Deploy the binary on a brand new VM" - what is the binary, and how do you deploy it, e.g. to a Digital Ocean droplet? I've been able to use docker for some apps, so if this is dockerized that would help somewhat. But, still hosting these things are beyond the capabilities of most people. Most people just want to be up and running and don't want to worry about setting anything up - and they will pay for it. Charge $25 a month for this and you will get tons of users.

bruh how do you not understand with deploy pocketbase ??? you just download the binary file and type serve in comman then app arealdy running

**Comment by benallfree** [comment]
For anyone following this thread, I have an alpha up at https://pockethost.io.

* Running on a single Digital Ocean VPS
* Frontend is Svelte plus a pocketbase backend (the  mothership)
* VPS has a daemon running that listens to new records from the main pocketbase instance. When it finds a new record, it provisions an SSL cert, updates nginx, creates a data storage location, finds a free port, and spawns a new pocketbase instance dedicated to the new entry. If anything ever crashes, this daemon will re-spawn all pocketbase instances at startup
* I keep an internal table (admin only) of instance info including health, uptime, port number, and other goodies
* Each pocketbase instance gets its own subdomain

Still to do:
* Is there a way to preconfigure the pocketbase instance with the correct Application URL?
* Is there a way to preconfigure the instance with a `<subdomain>@pockethost.io` email address? I have Amazon SES approved.
* It might make sense to shut down idle instances after 7 days or something. Waiting for a problem first.
* Still interested in a multitenant scenario where a single pocketbase instance can handle requests based on an `X-PocketBase-Tenant` header or something

@ganigeorgiev Let me know what you think. Also, if the PocketHost name is too close to PocketBase please tell me and I'll change it.

**Reply by ganigeorgiev** [reply]
Looks great. The UI may need a little more work, but overall I like how easy it is to start.

- > Is there a way to preconfigure the pocketbase instance with the correct Application URL?
  
  I'm not sure if you are asking for this, but extra init configuration like default address, port, etc. is in the roadmap and will be added somewhere after the refactoring.
  
- >  Is there a way to preconfigure the instance with a <subdomain>@pockethost.io email address? I have Amazon SES approved.
   
   Do you mean the mail settings? You can create a db migration and set the SMTP settings (currently there isn't too much documentation for that but I'll update it in the near future - https://pocketbase.io/docs/db-migrations/).

- > It might make sense to shut down idle instances after 7 days or something. Waiting for a problem first.
  
- > Still interested in a multitenant scenario where a single pocketbase instance can handle requests based on an X-PocketBase-Tenant header or something
  
  At least for v1.0.0 I don't think I'll have the time to work on this. I'm already behind my initial schedule for Presentator (the reason why PocketBase was created) and barely can handle the support with my day job.

About the name - it is ok, I don't mind. The domain is easy to type and pronounce so I think it is fine.

**Comment by benallfree** [comment]
PS, things I found very challenging that PocketHost solves or can solve...

* SMTP/outgoing email - very messy to set up if you have a custom domain. It's still not working right as I cannot receive mail at @pockethost.io even though the MX records are pointed to google.
* Server users, groups, and file permissions - always a problem :)
* Docker - a consistently discouraging experience for me, no matter how many times I try to use it
* SSL provisioning
* 8090->443 port mapping for a slightly cleaner experience
* Custom PocketBase builds

**Reply by jimafisk** [reply]
Cool project @benallfree! Is the code something you'd be willing to share for other folks who want to set up a PocketBase as a Service?

**Reply by benallfree** [reply]
@jimafisk I do have a repo started. There is a lot of un-Dockerized VPS config that is difficult to capture but I can easily share the site itself and the daemon that keeps everything organized. 

That’s funny though, Gani made a self-hosted tool that I turned into BaaS, now someone wants to self host that too!

Are you a Docker fan by chance? The project could really use containerization to capture the missing pieces.

**Reply by jimafisk** [reply]
Open source is a beautiful thing :blush:. Sure Docker is great!

**Reply by benallfree** [reply]
Cool, let me get the repo up and then perhaps we can collab!

**Reply by jimafisk** [reply]
Awesome thanks! No rush, I'm honestly a little overbooked in the short term, but think this is a cool idea :).

**Reply by benallfree** [reply]
@jimafisk Here you go https://github.com/benallfree/pockethost

**Comment by benallfree** [comment]
I wonder if node/Deno could be integrated with Go such that a hook written in JS could be called from Go. It would just need to receive and return enough context to keep things coherent. Beyond that, it could use the PocketBase JS client if it needed data access. I think JS hooks would be significantly easier than requiring users to learn Go.

**Reply by ganigeorgiev** [reply]
I don't plan implementing cloud function-style hooks, but collection triggers are in the roadmap - #317.
